packages/backend/src/nest/storage/communityMetadata/communityMetadata.store.ts

```diff
@@ -0,0 +1,217 @@
+    logger('Loaded community metadata to memory')
+    const meta = await this.getCommunityMetadata()
+    if (meta) {
+      emitter.emit(StorageEvents.COMMUNITY_METADATA_SAVED, meta)
+    }
+  }
+
+  public getAddress() {
+    return this.store?.address
+  }
+
+  public async close() {
+    await this.store?.close()
+  }
+
+  public async updateCommunityMetadata(newMeta: CommunityMetadata): Promise<CommunityMetadata | undefined> {
+    try {
+      // TODO: Also check OrbitDB identity when updating community metadata
+      const valid = await CommunityMetadataStore.validateCommunityMetadata(newMeta)
+      if (!valid) {
+        // TODO: Send validation errors to frontend or replicate
+        // validation on frontend?
+        logger.error('Failed to update community metadata')
+        return
+      }
+
+      logger(`About to update community metadata`)
+      if (!newMeta.id) return
+
+      // FIXME: update community metadata if it has changed (so that
+      // we can migrate community metadata easily)
+      const oldMeta = this.store.get(newMeta.id)
+      if (oldMeta?.ownerCertificate && oldMeta?.rootCa) {
+        return oldMeta
+      }
+
+      logger(`Updating community metadata`)
+      // @ts-expect-error - OrbitDB's type declaration of OrbitDB lacks identity
+      const ownerOrbitDbIdentity = this.orbitDb.identity.id
+      const meta = {
+        ...oldMeta,
+        ...newMeta,
+        ownerOrbitDbIdentity,
+      }
+
+      // putOwnerOrbitDbIdentity goes before store.put because the
+      // store's KeyValueIndex calls getOwnerOrbitDbIdentity
+      this.localDbService.putOwnerOrbitDbIdentity(ownerOrbitDbIdentity)
+
+      // FIXME: I think potentially there is a subtle developer
+      // experience bug here. Internally OrbitDB will call
+      // validateCommunityMetadataEntry and so validation may pass in
+      // this method, but still the entry is not added to the internal
+      // index. How can we detect that?
+      await this.store.put(meta.id, meta)
+
+      return meta
+    } catch (err) {
+      logger.error('Failed to add community metadata', err)
+    }
+  }
+
+  public static async validateCommunityMetadata(communityMetadata: CommunityMetadata): Promise<boolean> {
+    // FIXME: Add additional validation to verify communityMetadata
+    // contains required fields.
+
+    try {
+      // TODO: Should we sign community metadata with root private key
+      // and verify? I'm not sure it matters that much.
+
+      const rootCert = loadCertificate(communityMetadata.rootCa)
+      const ownerCert = loadCertificate(communityMetadata.ownerCertificate)
+
+      // Verify that owner certificate is signed by root certificate
+      return await ownerCert.verify(rootCert)
+    } catch (err) {
+      logger.error('Failed to validate community metadata:', communityMetadata.id, err?.message)
+      return false
+    }
+  }
+
+  public static async validateCommunityMetadataEntry(
```

Comment:
I don't think they would really see much yet. If they never received community metadata (or it was filtered out because it was invalid), the community name would be `...` I think and they wouldn't be able to validate certs from other users and the owner's name would be missing. We can further improve the UX at some point if desired. And perhaps the community owner can break users' ability to send/receive messages. I am not really sure. I think that depends on the rest of the codebase. Ideally it should only affect usernames/certificate validation. If somebody other than the owner attempts to edit community metadata, they would have to edit the Quiet codebase and depending on how much they edit there may or may not be an error. For everyone else, the entry will just get filtered out as if it didn't happen... there should be a log in the backend though that describes the failed validation.

Reply:
I made an issue to look into this later: https://github.com/orgs/TryQuiet/projects/3?pane=issue&itemId=46215134